ConfigurationBinder Bind to support scalar value, added new Get<T>

MichaCo opened the pull request:
+ New Get and Get<T> extension methods
+ Binding scalars works with all methods now in addition to POCOs.
+ New unit tests

#281

HaoK commented on src/Microsoft.Framework.Configuration.Binder/ConfigurationBinder.cs:
```diff
@@ -16,6 +16,34 @@ public static void Bind(this IConfiguration configuration, object model)
+
+        public static object Get(this IConfiguration configuration, Type type)
+        {
+            return BindInstance(type, null, configuration);
```
nit: you should use named parameters when passing null

HaoK commented on test/Microsoft.Framework.Configuration.Binder.Test/ConfigurationBinderTests.cs:
```diff
@@ -178,11 +221,17 @@ public void CanReadComplexProperties()
             var config = builder.Build();
-            var options = new ComplexOptions();
-            config.Bind(options);
+
```
Nit: its worth considering splitting these two tests BindCanReadComplexProperties / GetCanReadComplexProperties.  Slightly cleaner for this and the one below

HaoK commented:
Looks good to me, @divega thoughts?

divega commented:
Looks good in general. I have comments but I think they apply to the existing code even more than to the PR:
- I think we should decide on what behaviors we want for missing/null parameters. Then we should probably add some parameter checks based on that, but not necessarily everywhere, e.g. we might decide that in some scenarios the API will become a no-op if you don't pass some of them.
- I see some naming issues: 
 - For me model as a parameter name on the public API doesn't make much sense outside of MVC model binding.
 - As someone looking at the code to try to understand it is hard to see what the difference is between `BindObject()` and `BindInstance()`. This one is not public API so it is just internal code quality.
 - I am not sure why `BindInstance()` ignores the passed object some times. It might intentional and an implementation detail but it smells to me.
- I have been looking at how this API could be used and I think it is interesting to consider the possibility of allowing passing default values, e.g.:

``` C#
var timeout = config.Get<int>("timeout", defaultValue: 30);
```

Some users would prefer to add those values in an in-memory configuration source so that they always get returned, but the pattern above seems to be less hassle.

divega commented:
And regarding the last bullet on default values, what happens now if I ask for e.g. an int and:

1. The value found is null? 
2. The value found is not null but cannot be converted to int?

MichaCo commented:
@divega I can rename `model` to `obj` for bind, any better idea for the name? ;)

BindObject -> I thought its the best name because it is only binding all properties of the object which gets passed in.

Regarding BindInstance, it ignores the object in case the method should return the value of a section already. And if the object passed in is null, it tries to instantiate it and then calls the different bind methods. This method does a lot of different stuff, maybe it could be refactored into multiple methods, or we just rename it. Any idea?

Regarding the default values, I agree that this would be nice, but I guess this should be another PR and Task, right?

divega commented:
> I can rename  model  to  obj  for bind, any better idea for the name? ;)

I would suggest using `instance` for this. It is practically a synonym of object, which is why it is so confusing to have a method called `BingObject()` and another one called `BindInstance()` in the implementation :smile:

> BindObject 

What about `BindNonScalar` or `BindStructuralObject`. Anything to convey that this is for the path in which we didn't bind a value.

> BindInstance ... refactored

Don't worry about that for now.

> default values... 

Agreed it should be a different PR. I will file a bunch of new bugs for these and some other concerns I have.

divega commented:
Filed #285 and #286. @HaoK can you do a final review to see if this is ready?

MichaCo commented:
@HaoK I fixed your PR comments and also added a few more unit tests
@divega 

> what happens now if I ask for e.g. an int and 
> The value found is null?

I actually found a bug in `Get<T>` when `BindInstance` returns null, the cast to `(T)` of course throws a null reference exception. This is fixed now and has a unit test.

In general, if the value found is null, we will return `default(T)`, [see unit tests](https://github.com/aspnet/Configuration/pull/282/files#diff-f162abdb45ea855d1b299c1d26c0f385R95).
That being said, in case of collections, the value will not be added to the list! [see unit tests](https://github.com/aspnet/Configuration/pull/282/files#diff-5678dfc08305f4141744e5b25329faacR38)

> The value found is not null but cannot be converted to int?

That's already being tested, `Get` will throw the same exception as `Bind` did before all the changes.

There is still one possible null reference exception in the code which is if the `configuration` parameter on `Bind` or `Get/Get<T>` is null. This could happen if someone calls `ConfigurationBinder.Bind` directly. In this case, `BindInstance` will throw the exception...
I think we should check for null in `BindInstance` and return the passed in instance/null. The same happens if `config.GetChildren().Any()` is false anyways.

divega commented:
>I think we should check for null in BindInstance and return the passed in instance/null. The same happens if config.`GetChildren().Any()` is false anyways.

Sounds good, however I would expect the public methods to throw if a null IConfiguration is passed at the start. Thoughts?

MichaCo commented:
Yup makes sense, should I add the [NotNull] attribute or check for null and throw an ArgumentException directly?

divega commented:
Both. We don't know if we'll get to finish the code that auto-generates the checks based on the attribute.

MichaCo commented:
Ok that's done, too ;)

HaoK commented:
Re [NotNull] my preference is to just use the attribute for new code which seems to be the pattern in most places.

divega commented:
@HaoK actually you need to change your preference :smile:

divega commented on src/Microsoft.Framework.Configuration.Binder/ConfigurationBinder.cs:
```diff
@@ -7,18 +7,66 @@
+            if (configuration == null)
             {
-                foreach (var property in GetAllProperties(model.GetType().GetTypeInfo()))
+                throw new ArgumentNullException("configuration");
```
Use nameof(configuration)

divega commented on src/Microsoft.Framework.Configuration.Binder/ConfigurationBinder.cs:
```diff
@@ -7,18 +7,66 @@
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException("configuration");
```
`nameof(configuration)`

divega commented:
@HaoK was right re `[NotNull]`. It seems we are not using it consistently in this repo so there is very low value in adding it for this methods. On the other hand it doesn't hurt so no need to remove it from the PR either.

In general LGTM with minor feedback. @HaoK want to do another last review? :smile:

MichaCo commented:
@divega You comments are fixed, too

divega commented:
LGTM. @HaoK please finish reviewing and take this. We might need to rebase with https://github.com/aspnet/Configuration/pull/288/.

HaoK commented:
Only minor thing is can we nuke the [NotNull] since we are removing them everywhere, lets not add more now

MichaCo commented:
I removed the [NotNull] and squashed everything into one commit. Didn't merge yet, let me know if you want me to do that

HaoK commented:
Looks good to me, merge away

divega commented:
Thanks for your contribution @MichaCo!

HaoK commented:
Hrm now we throw on Core:

Microsoft.Framework.Configuration.Binder.Test.ConfigurationBinderTests.GetUri [FAIL]
      System.InvalidOperationException : Failed to convert 'http://www.bing.com' to type 'System.Uri'.

 Microsoft.Framework.Configuration.Binder.Test.ConfigurationBinderTests.CanReadAllSupportedTypes(value: "http://www.bing.com", type: typeof(System.Uri)) [FAIL]
      System.NotSupportedException : TypeConverter cannot convert from System.String.

HaoK commented:
I'll just disable these two new tests on core since it doesn't appear to be supported

MichaCo commented:
@divega welcome ;)
@HaoK oh wow sorry about that! I was pretty sure I ran tests on both clr and coreclr.
Maybe you should add an issue to the https://github.com/dotnet/corefx repo?
